<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accurate identification of co-methylated and differentially methylated
  regions in epigenome-wide association studies • coMethDMR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Accurate identification of co-methylated and differentially methylated
  regions in epigenome-wide association studies">
<meta property="og:description" content="coMethDMR identifies genomic regions associated with continuous
  phenotypes by optimally leverages covariations among CpGs within predefined
  genomic regions. Instead of testing all CpGs within a genomic region,
  coMethDMR carries out an additional step that selects co-methylated
  sub-regions first without using any outcome information. Next, coMethDMR tests
  association between methylation within the sub-region and continuous phenotype
  using a random coefficient mixed effects model, which models both variations
  between CpG sites within the region and differential methylation
  simultaneously. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">coMethDMR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vin1_Introduction_to_coMethDMR_geneBasedPipeline.html">Introduction to coMethDMR</a>
    </li>
    <li>
      <a href="articles/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.html">Genome-wide methylation analysis using coMethDMR via parallel computing</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TransBioInfoLab/coMethDMR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level2">
<h2 id="comethdmr-accurate-identification-of-co-methylated-and-differentially-methylated-regions-in-epigenome-wide-association-studies">coMethDMR: Accurate identification of co-methylated and differentially methylated regions in epigenome-wide association studies<a class="anchor" aria-label="anchor" href="#comethdmr-accurate-identification-of-co-methylated-and-differentially-methylated-regions-in-epigenome-wide-association-studies"></a>
</h2>
<p>Gomez L, Odom GJ, Young JI, Martin ER, Liu L, Chen X, Griswold AJ, Gao Z, Zhang L, Wang L (2019) Nucleic Acids Research, gkz590, <a href="https://doi.org/10.1093/nar/gkz590" class="external-link uri">https://doi.org/10.1093/nar/gkz590</a></p>
</div>
<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p>coMethDMR is an R package that identifies genomic regions associated with continuous phenotypes by optimally leverages covariations among CpGs within predefined genomic regions. Instead of testing all CpGs within a genomic region, coMethDMR carries out an additional step that selects comethylated sub-regions first without using any outcome information. Next, coMethDMR tests association between methylation within the sub-region and continuous phenotype using a random coefficient mixed effects model, which models both variations between CpG sites within the region and differential methylation simultaneously.</p>
</div>
<div class="section level2">
<h2 id="quick-overview--analysis-steps">Quick Overview / Analysis Steps<a class="anchor" aria-label="anchor" href="#quick-overview--analysis-steps"></a>
</h2>
<p>Assuming you have completed all pre-processing and normalization procedures, here are the steps to analyse your Illumina EPIC or 450k DNA methylation data.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Your Data</strong>:
<ul>
<li>Ensure your methylation data is loaded into R’s Global Environment as a numeric matrix in <em>probe by sample</em> form: with probe IDs as your row names and sample IDs as the column names</li>
<li>Your phenotype / clinical data should be a data frame with a column called <code>Sample</code> (spelled exactly with an upper-case “S”); these sample IDs should match the column names of the methylation data</li>
</ul>
</li>
<li>
<strong>Pre-Defined Regions</strong>: Load the list of pre-calculated regions of “contiguous” CpGs which matches your Illumina data type. We have pre-calculated some of these lists of regions. We used the <code><a href="reference/CloseBySingleRegion.html">CloseBySingleRegion()</a></code> function with <code>maxGap = 200</code> (genomic locations within 200 base pairs are placed in the same cluster) and <code>minCpGs = 3</code> (we need at least 3 CpGs to retain the location). These data files are:
<ul>
<li>Genic regions, 450k array: <code>extdata/450k_Gene_3_200.rds</code>. Load this via <code>system.file("extdata", "450k_Gene_3_200.rds", package = "coMethDMR", mustWork = TRUE)</code>
</li>
<li>Inter-genic regions, 450k array: <code>extdata/450k_InterGene_3_200.rds</code>.Load this via <code>system.file("extdata", "450k_InterGene_3_200.rds", package = "coMethDMR", mustWork = TRUE)</code>
</li>
<li>Genic regions, EPIC array: download the supplemental data file from <a href="https://github.com/TransBioInfoLab/coMethDMR_data/blob/main/data/EPIC_10b4_Gene_3_200.rds" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_data/blob/main/data/EPIC_10b4_Gene_3_200.rds</a>
</li>
<li>Inter-genic regions, EPIC array: download the supplemental data file from <a href="https://github.com/TransBioInfoLab/coMethDMR_data/blob/main/data/EPIC_10b4_InterGene_3_200.rds" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_data/blob/main/data/EPIC_10b4_InterGene_3_200.rds</a>
</li>
</ul>
</li>
<li>
<strong>Adjust Methylation for Covariates</strong> with the <code><a href="reference/GetResiduals.html">GetResiduals()</a></code> function; your methylation values may be confounded by clinical variables unrelated to your treatment, such as sex, age, or even <a href="https://www.nature.com/articles/s41598-021-88504-0" class="external-link">the square of age</a>
<ul>
<li>
<em>Input</em>: your methylation data and covariates from <strong>Step 1</strong>
</li>
<li>
<em>Output</em>: a matrix of methylation residuals in <em>probe by sample</em> form</li>
</ul>
</li>
<li>
<strong>Regions of Co-Methylation</strong> with the <code><a href="reference/CoMethAllRegions.html">CoMethAllRegions()</a></code> function
<ul>
<li>
<em>Input</em>: the list of regions of interest from <strong>Step 2</strong>; the methylation residuals (or the unadjusted M-values) from <strong>Step 3</strong>
</li>
<li>
<em>Output</em>: either a list of data frames (if <code>output = "dataframe"</code>) or a list of vectors of probe IDs (if <code>output = "CpGs"</code>)
<ul>
<li>Data Frame: one data frame per region of concurrent methylation; each data frame has a row per probe and columns for the probe ID, it’s enclosing region, chromosome, correlation with the surrounding probes, and an indicator for “co”-methylation</li>
<li>Vector of CpGs: one vector per region of concurrent methylation; the vector simply contains the probe IDs of the CpGs in each region</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>Detecting Regions Differential Co-Methylation</strong> with <code><a href="reference/lmmTestAllRegions.html">lmmTestAllRegions()</a></code>; this function performs a linear mixed model test to detect, from regions of concurrent methylation, which regions have differential methylation across the majority of probes (so that one single probe cannot influence the entire region)
<ul>
<li>
<em>Input</em>: the methylation and clinical data from <strong>Step 1</strong> and the regions of concurrent methylation from <strong>Step 4</strong>
</li>
<li>
<em>Output</em>: a data frame with
<ul>
<li>one row per region, and</li>
<li>columns for the region chromosome and range, the number of CpGs in the region, and the linear mixed model summaries for that region (including regression slope estimate, standard error, test statistic, <span class="math inline"><em>p</em></span>-value, and false discovery rate)</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>Annotate Statistically Significant Regions</strong> with <code><a href="reference/AnnotateResults.html">AnnotateResults()</a></code>; this function takes the linear mixed model summaries by region from <strong>Step 5</strong> and adds on reference gene symbols and their relationships to known CpG islands</li>
</ol>
<p>See our</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="bioconductor-version">Bioconductor Version<a class="anchor" aria-label="anchor" href="#bioconductor-version"></a>
</h3>
<p>The <code>coMethDMR::</code> package has been accepted to the <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> repository of R packages. It will be included in version 3.15 (April 2022 release). To install <a href="https://www.bioconductor.org/packages/devel/bioc/html/coMethDMR.html" class="external-link">this version</a>, please use the following code:</p>
<pre><code><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="co"># The following initializes usage of Bioc devel</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version<span class="op">=</span><span class="st">'devel'</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"coMethDMR"</span><span class="op">)</span></code></pre>
</div>
<div class="section level3">
<h3 id="development-version">Development Version<a class="anchor" aria-label="anchor" href="#development-version"></a>
</h3>
<p>The development version of <code>coMethDMR::</code> can also be installed from this GitHub repository by</p>
<p><code>{r eval=FALSE, message=FALSE, warning=FALSE, results='hide'} library(devtools) install_github("TransBioInfoLab/coMethDMR")</code></p>
<p>Please note that using compiled code from GitHub may require your computer to have additional software (<a href="https://cran.r-project.org/bin/windows/Rtools/rtools40.html" class="external-link">Rtools</a> for Windows or <a href="https://developer.apple.com/xcode/" class="external-link">Xcode</a> for Mac). Also note that installing this development version may result in some errors. We have outlined potential troubleshooting steps below.</p>
<div class="section level4">
<h4 id="install-errors-cache">Install Errors: Cache<a class="anchor" aria-label="anchor" href="#install-errors-cache"></a>
</h4>
<p>You may get the following error during installation:</p>
<pre><code>Error: package or namespace load failed for 'coMethDMR':
 .onLoad failed in loadNamespace() for 'coMethDMR', details:
  call: .updateHubDB(hub_bfc, .class, url, proxy, localHub)
  error: Invalid Cache: sqlite file
  Hub has not been added to cache
  Run again with 'localHub=FALSE'
Error: loading failed</code></pre>
<p>If so, please fix this by running <code><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub::ExperimentHub()</a></code> first (and type <code>yes</code> if you receive a prompt to create a local cache for your data), then re-installing the package. Please see this white paper for more information: <a href="https://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/TroubleshootingTheCache.html" class="external-link uri">https://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/TroubleshootingTheCache.html</a>.</p>
</div>
<div class="section level4">
<h4 id="install-errors-onload-failure">Install Errors: <code>.onLoad()</code> Failure<a class="anchor" aria-label="anchor" href="#install-errors-onload-failure"></a>
</h4>
<p>You may also get this error during installation:</p>
<pre><code>Error: package or namespace load failed for 'coMethDMR':
 .onLoad failed in loadNamespace() for 'coMethDMR', details:
  call: NULL
  error: $ operator is invalid for atomic vectors</code></pre>
<p>This error is caused by a version mismatch issue for the <code>sesameData::</code> (<a href="https://bioconductor.org/packages/sesameData/" class="external-link uri">https://bioconductor.org/packages/sesameData/</a>) package. We require <code>sesameData::</code> version 1.12 or higher. To fix this, you will need Biocdonductor version 3.14 or later. The following code will assist here:</p>
<pre><code><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"sesameData"</span><span class="op">)</span></code></pre>
<p>After successfully executing the above installation, you should be able to install <code>coMethDMR::</code> from GitHub like normal.</p>
</div>
</div>
<div class="section level3">
<h3 id="loading-the-package">Loading the Package<a class="anchor" aria-label="anchor" href="#loading-the-package"></a>
</h3>
<p>After installation, the coMethDMR package can be loaded into R using:</p>
<p><code>{r eval=TRUE, message=FALSE, warning=FALSE, results='hide'} library(coMethDMR)</code></p>
</div>
</div>
<div class="section level2">
<h2 id="manual">Manual<a class="anchor" aria-label="anchor" href="#manual"></a>
</h2>
<p>The reference manual for coMethDMR can be downloaded from old repository: <a href="https://github.com/TransBioInfoLab/coMethDMR_old/tree/master/docs/" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_old/tree/master/docs/</a>. The reference manual is <a href="https://github.com/TransBioInfoLab/coMethDMR_old/blob/master/docs/coMethDMR_0.0.0.9001.pdf" class="external-link">coMethDMR_0.0.0.9001.pdf</a>. Two vignettes are available in the same directory: <a href="https://github.com/TransBioInfoLab/coMethDMR_old/blob/master/docs/1_Introduction_coMethDMR_10-9-2019.pdf" class="external-link">1_Introduction_coMethDMR_10-9-2019.pdf</a> and <a href="https://github.com/TransBioInfoLab/coMethDMR_old/blob/master/docs/2_BiocParallel_for_coMethDMR_geneBasedPipeline.pdf" class="external-link">2_BiocParallel_for_coMethDMR_geneBasedPipeline.pdf</a>.</p>
</div>
<div class="section level2">
<h2 id="frequently-asked-questions">Frequently Asked Questions<a class="anchor" aria-label="anchor" href="#frequently-asked-questions"></a>
</h2>
<ol style="list-style-type: decimal">
<li>There are two main steps in coMethDMR: (1) identifying comethylatyed clusters (2) testing methylation levels in those comethylated clusters against a phenotype”. In step 1 and 2, should we use beta value or M values for CoMethDMR?</li>
</ol>
<p>Answer: In step (1), using M values and beta values produce similar results. See Supplementary Table 2 Comparison of using beta values or M-values for identifying co-methylated regions in first step of coMethDMR pipeline at optimal rdrop parameter value of the coMethDMR paper.</p>
<p>In step (2), M-values should be used because it has better statistical properties. See Du et al. (2010) Comparison of Beta-value and M-value methods for quantifying methylation levels by microarray analysis.</p>
</div>
<div class="section level2">
<h2 id="development-history">Development History<a class="anchor" aria-label="anchor" href="#development-history"></a>
</h2>
<p>Our development history is at <a href="https://github.com/TransBioInfoLab/coMethDMR_old" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_old</a></p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/coMethDMR" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/TransBioInfoLab/coMethDMR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/TransBioInfoLab/coMethDMR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing coMethDMR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Fernanda Veitzman <br><small class="roles"> Maintainer </small>  </li>
<li>Lissette Gomez <br><small class="roles"> Author </small>  </li>
<li>Tiago Silva <br><small class="roles"> Author </small>  </li>
<li>Lily Wang <br><small class="roles"> Author </small>  </li>
<li>Gabriel Odom <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Fernanda Veitzman, Lissette Gomez, Tiago Silva, Lily Wang, Gabriel Odom.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
