<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to coMethDMR • coMethDMR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to coMethDMR">
<meta property="og:description" content="coMethDMR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coMethDMR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vin1_Introduction_to_coMethDMR_geneBasedPipeline.html">Introduction to coMethDMR</a>
    </li>
    <li>
      <a href="../articles/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.html">Genome-wide methylation analysis using coMethDMR via parallel computing</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TransBioInfoLab/coMethDMR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to coMethDMR</h1>
                        <h4 data-toc-skip class="author">Lissette Gomez, Gabriel Odom, Tiago Chedraoui Silva, Lanyu Zhang, and Lily Wang</h4>
            
            <h4 data-toc-skip class="date">March 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TransBioInfoLab/coMethDMR/blob/HEAD/vignettes/vin1_Introduction_to_coMethDMR_geneBasedPipeline.Rmd" class="external-link"><code>vignettes/vin1_Introduction_to_coMethDMR_geneBasedPipeline.Rmd</code></a></small>
      <div class="hidden name"><code>vin1_Introduction_to_coMethDMR_geneBasedPipeline.Rmd</code></div>

    </div>

    
    
<p><code>coMethDMR</code> is an R package that identifies genomic regions that are both co-methylated and differentially methylated in Illumina array datasets. Instead of testing all CpGs within a genomic region, <code>coMethDMR</code> carries out an additional step that selects co-methylated sub-regions first without using any outcome information. Next, <code>coMethDMR</code> tests association between methylation within the sub-region and continuous phenotype using a random coefficient mixed effects model, which models both variations between CpG sites within the region and differential methylation simultaneously.</p>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The latest version can be installed by</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"coMethDMR"</span><span class="op">)</span></code></pre></div>
<p>After installation, the <code>coMethDMR</code> package can be loaded into R using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TransBioInfoLab/coMethDMR" class="external-link">coMethDMR</a></span><span class="op">)</span></code></pre></div>
<p>If you are running the <code>coMethDMR</code> package for the first time, you will also need to download supplemental data sets from the <code>sesameData</code> package. This happens automatically for the 450k and EPIC arrays when you load <code>coMethDMR</code> for the first time. If you experience errors in this process, you may have your cache in an older default location. Please follow these steps to update where these genomic data are stored: <a href="https://bioconductor.org/packages/devel/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html#default-caching-location-update" class="external-link uri">https://bioconductor.org/packages/devel/bioc/vignettes/ExperimentHub/inst/doc/ExperimentHub.html#default-caching-location-update</a>.</p>
</div>
<div class="section level3">
<h3 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a>
</h3>
<div class="section level4">
<h4 id="example-methylation-data">Example Methylation Data<a class="anchor" aria-label="anchor" href="#example-methylation-data"></a>
</h4>
<p>The input of <code>coMethDMR</code> are methylation beta values. We assume quality control and normalization of the methylation dataset have been performed, by R packages such as <code>minfi</code> or <code>RnBeads</code>. For illustration, we use a subset of prefrontal cortex methylation data (GEO GSE59685) from a recent Alzheimer’s disease epigenome-wide association study which was described in Lunnon et al. (2014). This example dataset contains beta values for 8552 CpGs on chromosome 22 for a random selection of 20 subjects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span>
<span class="va">betasChr22_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##            GSM1443279 GSM1443663 GSM1443434 GSM1443547 GSM1443577</span>
<span class="co">## cg00004192  0.9249942  0.8463296  0.8700718  0.9058205  0.9090382</span>
<span class="co">## cg00004775  0.6523025  0.6247554  0.7573476  0.6590817  0.6726261</span>
<span class="co">## cg00012194  0.8676339  0.8679048  0.8484754  0.8754985  0.8484458</span>
<span class="co">## cg00013618  0.9466056  0.9475467  0.9566493  0.9588431  0.9419563</span>
<span class="co">## cg00014104  0.3932388  0.5525716  0.4075900  0.3997278  0.3216956</span></code></pre>
<p>If you have used either the <a href="https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html" class="external-link">minfi</a> or <a href="https://bioconductor.org/packages/release/bioc/vignettes/RnBeads/inst/doc/RnBeads.pdf" class="external-link">RnBeads</a> packages to pre-process your methylation data, we now show a quick example of the code necessary to create a methylation data frame similar to <code>betasChr22_df</code> above (note that <code>dataObject_minfi</code> and <code>dataObject_RnBeads</code> are objects containing the pre-processed DNA methylation data as returned by the <code>minfi::</code> or <code>RnBeads::</code> packages, respectively):</p>
<pre><code><span class="co">###  minfi  ###</span>
<span class="va">betas_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span>
  <span class="fu">minfi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/minfi/man/utils.html" class="external-link">getMethSignal</a></span><span class="op">(</span><span class="va">dataObject_minfi</span>, what <span class="op">=</span> <span class="st">"Beta"</span><span class="op">)</span>
<span class="op">)</span>

<span class="co">###  RnBeads  ###</span>
<span class="va">betas_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span>
  <span class="fu">RnBeads</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RnBeads/man/meth-methods.html" class="external-link">meth</a></span><span class="op">(</span><span class="va">dataObject_RnBeads</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">)</span></code></pre>
</div>
<div class="section level4">
<h4 id="example-response-data">Example Response Data<a class="anchor" aria-label="anchor" href="#example-response-data"></a>
</h4>
<p>The corresponding phenotype dataset included variables <code>stage</code> (Braak AD stage), <code>subject.id</code>, <code>slide</code> (batch effect), <code>Sex</code>, <code>Sample</code> and <code>age.brain</code> (age of the brain donor). Please note the phenotype file needs to have a variable called “Sample” that will be used by coMethDMR to link to the methylation dataset.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pheno_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno_df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    stage subject.id         sex     Sample age.brain       slide</span>
<span class="co">## 3      0          1 Sex: FEMALE GSM1443251        82  6042316048</span>
<span class="co">## 8      2          2 Sex: FEMALE GSM1443256        82  6042316066</span>
<span class="co">## 10    NA          3   Sex: MALE GSM1443258        89  6042316066</span>
<span class="co">## 15     1          4 Sex: FEMALE GSM1443263        81  7786923107</span>
<span class="co">## 21     2          5 Sex: FEMALE GSM1443269        92  6042316121</span>
<span class="co">## 22     1          6   Sex: MALE GSM1443270        78  6042316099</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="a-quick-work-through-of-comethdmr">A quick work through of coMethDMR<a class="anchor" aria-label="anchor" href="#a-quick-work-through-of-comethdmr"></a>
</h3>
<p>For illustration, suppose we are interested in identifying co-methylated genomic regions associated with AD stages (<code>stage</code> treated as a linear variable). Here we demonstrate analysis of genomic regions mapped to CpG islands on chromosome 22. However the workflow can be similarly conducted for other types of genomic regions. See details in Section 2.1 below for gene based pipeline that tests genic and intergenic regions.</p>
<p>There are several steps: (1) obtain CpGs located closely (see details in Section 2.1 below) in genomic regions mapped to CpG islands, (2) identify co-methylated regions, and (3) test co-methylated regions against the outcome variable AD stage.</p>
<p>For the first step, we use the following commands:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CpGisland_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"CpGislandsChr22_ex.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>Here, <code>CpGisland_ls</code> is a list of 20 items, with each item of the list including a group of CpG probe IDs located closely within a particular CpG island region. Section 2.1 discusses how to import additional types of genomic regions.</p>
<p>Next, we identify co-methylated regions based on Mvalues.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span>
  <span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
    dnam <span class="op">=</span> <span class="va">betasChr22_df</span>,
    betaToM <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># converts beta to m-values</span>
    method <span class="op">=</span> <span class="st">"pearson"</span>,
    CpGs_ls <span class="op">=</span> <span class="va">CpGisland_ls</span>,
    arrayType <span class="op">=</span> <span class="st">"450k"</span>,
    returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>, 
    output <span class="op">=</span> <span class="st">"CpGs"</span>,
    nCores_int <span class="op">=</span> <span class="fl">1</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">##    user  system elapsed </span>
<span class="co">##   13.02    1.67   15.27</span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ~15 seconds</span>

<span class="va">coMeth_ls</span></code></pre></div>
<pre><code><span class="co">## $`chr22:18268062-18268249`</span>
<span class="co">## [1] "cg12460175" "cg14086922" "cg21463605"</span>
<span class="co">## </span>
<span class="co">## $`chr22:18324579-18324769`</span>
<span class="co">## [1] "cg19606103" "cg14031491" "cg03816851"</span>
<span class="co">## </span>
<span class="co">## $`chr22:18531243-18531447`</span>
<span class="co">## [1] "cg25257671" "cg06961233" "cg08819022"</span></code></pre>
<p><code>coMeth_ls</code> is list with that contains groups of CpG probeIDs corresponding to co-methylated regions. Three comethylated regions were identified in this example.</p>
<p>If we want to look at co-methylation within the first co-methylated region:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">WriteCorrPlot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">beta_mat</span><span class="op">)</span><span class="op">{</span>

  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/taiyun/corrplot" class="external-link">corrplot</a></span><span class="op">)</span>
  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/TransBioInfoLab/coMethDMR" class="external-link">coMethDMR</a></span><span class="op">)</span>

  <span class="va">CpGs_char</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">beta_mat</span><span class="op">)</span>

  <span class="va">CpGsOrd_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/OrderCpGsByLocation.html">OrderCpGsByLocation</a></span><span class="op">(</span>
    <span class="va">CpGs_char</span>, arrayType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"450k"</span><span class="op">)</span>, output <span class="op">=</span> <span class="st">"dataframe"</span>
  <span class="op">)</span>

  <span class="va">betaOrdered_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">beta_mat</span><span class="op">[</span><span class="va">CpGsOrd_df</span><span class="op">$</span><span class="va">cpg</span> ,<span class="op">]</span><span class="op">)</span>

  <span class="va">corr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span>
    <span class="va">betaOrdered_mat</span>, method <span class="op">=</span> <span class="st">"spearman"</span>, use <span class="op">=</span> <span class="st">"pairwise.complete.obs"</span>
  <span class="op">)</span>

  <span class="fu"><a href="https://rdrr.io/pkg/corrplot/man/corrplot.html" class="external-link">corrplot</a></span><span class="op">(</span><span class="va">corr</span>, method <span class="op">=</span> <span class="st">"number"</span>, number.cex <span class="op">=</span> <span class="fl">1</span>, tl.cex <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># subsetting beta values to include only co-methylated probes</span>
<span class="va">betas_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span>
  <span class="va">betasChr22_df</span>, 
  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">coMeth_ls</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>
<span class="op">)</span>

<span class="fu">WriteCorrPlot</span><span class="op">(</span><span class="va">betas_df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Warning: package 'corrplot' was built under R version 4.1.2</span></code></pre>
<p><img src="vin1_Introduction_to_coMethDMR_geneBasedPipeline_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Next, we test these co-methylated regions against <code>stage</code> using a random coefficient model (more details in section 2.3 below).</p>
<p>Some messages are generated during mixed models fitting, which are saved to the file specified by <code>outLogFile</code>. The interpretations of these messages can be found in the FAQs at the end of this document (see Section 3, item (1) and (2)).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmTestAllRegions.html">lmmTestAllRegions</a></span><span class="op">(</span>
  betas <span class="op">=</span> <span class="va">betasChr22_df</span>,
  region_ls <span class="op">=</span> <span class="va">coMeth_ls</span>,
  <span class="va">pheno_df</span>,
  contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
  covariates_char <span class="op">=</span> <span class="cn">NULL</span>,
  modelType <span class="op">=</span> <span class="st">"randCoef"</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>
  <span class="co"># generates a log file in the current directory</span>
  <span class="co"># outLogFile = paste0("lmmLog_", Sys.Date(), ".txt")</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## Analyzing region chr22:18268062-18268249.</span></code></pre>
<pre><code><span class="co">## Analyzing region chr22:18324579-18324769.</span></code></pre>
<pre><code><span class="co">## Analyzing region chr22:18531243-18531447.</span></code></pre>
<pre><code><span class="co">## For future calls to this function, perhaps specify a log file.</span>
<span class="co">##       Set the file name of the log file with the outLogFile argument.</span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out_df</span></code></pre></div>
<pre><code><span class="co">##   chrom    start      end nCpGs    Estimate     StdErr      Stat     pValue</span>
<span class="co">## 1 chr22 18268062 18268249     3 -0.06678558 0.03884685 -1.719202 0.08557761</span>
<span class="co">## 2 chr22 18324579 18324769     3  0.03549924 0.02885584  1.230227 0.21861213</span>
<span class="co">## 3 chr22 18531243 18531447     3 -0.05181161 0.05082589 -1.019394 0.30801599</span>
<span class="co">##         FDR</span>
<span class="co">## 1 0.2567328</span>
<span class="co">## 2 0.3080160</span>
<span class="co">## 3 0.3080160</span></code></pre>
<p>Here <code>out_df</code> is a data frame of genomic regions, with corresponding p-values and false discovery rate (FDRs) from the random coefficient mixed model.</p>
<p>We can annotate these results by adding corresponding genes and probes mapped to the genomic regions.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span>
  <span class="va">outAnno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AnnotateResults.html">AnnotateResults</a></span><span class="op">(</span>
    lmmRes_df <span class="op">=</span> <span class="va">out_df</span>,
    arrayType <span class="op">=</span> <span class="st">"450k"</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    user  system elapsed </span>
<span class="co">##    7.66    1.33    9.00</span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># ~12 seconds</span>

<span class="va">outAnno_df</span></code></pre></div>
<pre><code><span class="co">##   chrom    start      end nCpGs    Estimate     StdErr      Stat     pValue</span>
<span class="co">## 1 chr22 18268062 18268249     3 -0.06678558 0.03884685 -1.719202 0.08557761</span>
<span class="co">## 2 chr22 18324579 18324769     3  0.03549924 0.02885584  1.230227 0.21861213</span>
<span class="co">## 3 chr22 18531243 18531447     3 -0.05181161 0.05082589 -1.019394 0.30801599</span>
<span class="co">##         FDR UCSC_RefGene_Group UCSC_RefGene_Accession UCSC_RefGene_Name</span>
<span class="co">## 1 0.2567328                                                            </span>
<span class="co">## 2 0.3080160               Body NM_001122731;NM_015241            MICAL3</span>
<span class="co">## 3 0.3080160                                                            </span>
<span class="co">##   Relation_to_Island</span>
<span class="co">## 1             Island</span>
<span class="co">## 2             Island</span>
<span class="co">## 3             Island</span></code></pre>
<p>To further examine the significant regions, we can also extract individual CpG p-values within these significant regions. For example, for the most significant region <code>chr22:18268062-18268249</code>,</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outCpGs_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CpGsInfoOneRegion.html">CpGsInfoOneRegion</a></span><span class="op">(</span>
 regionName_char <span class="op">=</span> <span class="st">"chr22:18268062-18268249"</span>,
 betas_df <span class="op">=</span> <span class="va">betasChr22_df</span>,
 pheno_df <span class="op">=</span> <span class="va">pheno_df</span>,
 contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
 covariates_char <span class="op">=</span> <span class="cn">NULL</span>,
 arrayType <span class="op">=</span> <span class="st">"450k"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outCpGs_df</span></code></pre></div>
<pre><code><span class="co">##                    Region        cpg   chr      pos slopeEstimate slopePval</span>
<span class="co">## 1 chr22:18268062-18268249 cg12460175 chr22 18268062       -0.0329    0.3785</span>
<span class="co">## 2 chr22:18268062-18268249 cg14086922 chr22 18268239       -0.0724    0.0667</span>
<span class="co">## 3 chr22:18268062-18268249 cg21463605 chr22 18268249       -0.0951    0.0255</span>
<span class="co">##   UCSC_RefGene_Name UCSC_RefGene_Accession UCSC_RefGene_Group</span>
<span class="co">## 1                                                            </span>
<span class="co">## 2                                                            </span>
<span class="co">## 3</span></code></pre>
<p>These CpGs mapped to intergenic regions, so there are no gene names associated with the probes. For genic regions such as <code>chr22:19709548-19709755</code>, we would have results such as the following:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># library("GenoGAM")</span>
<span class="co"># NOTE 2022-03-28: what are we using this package for again?</span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/CpGsInfoOneRegion.html">CpGsInfoOneRegion</a></span><span class="op">(</span>
  regionName_char <span class="op">=</span> <span class="st">"chr22:19709548-19709755"</span>,
  betas_df <span class="op">=</span> <span class="va">betasChr22_df</span>,
  pheno_df <span class="op">=</span> <span class="va">pheno_df</span>,
  contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
  covariates_char <span class="op">=</span> <span class="cn">NULL</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">##                    Region        cpg   chr      pos slopeEstimate slopePval</span>
<span class="co">## 1 chr22:19709548-19709755 cg04533276 chr22 19709548       -0.0656    0.1529</span>
<span class="co">## 2 chr22:19709548-19709755 cg20193802 chr22 19709696       -0.0346    0.4808</span>
<span class="co">## 3 chr22:19709548-19709755 cg05726109 chr22 19709755        0.0021    0.9585</span>
<span class="co">##   UCSC_RefGene_Name UCSC_RefGene_Accession UCSC_RefGene_Group</span>
<span class="co">## 1             SEPT5              NM_002688               Body</span>
<span class="co">## 2       SEPT5;GP1BB    NM_002688;NM_000407       Body;TSS1500</span>
<span class="co">## 3       SEPT5;GP1BB    NM_002688;NM_000407       Body;TSS1500</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="details-of-comethdmr-workflow">Details of <code>coMethDMR</code> workflow<a class="anchor" aria-label="anchor" href="#details-of-comethdmr-workflow"></a>
</h2>
<div class="section level3">
<h3 id="genomic-regions-tested-in-gene-based-pipeline">Genomic regions tested in gene based pipeline<a class="anchor" aria-label="anchor" href="#genomic-regions-tested-in-gene-based-pipeline"></a>
</h3>
<p>Genomic regions on the Illumina arrays can be defined based on their relations to genes or CpG Islands. To reduce redundancy in the tested genomic regions, we recommend first testing genic and intergenic regions, then add annotations to each genomic region for their relation to CpG islands.</p>
<p>In <code>coMethDMR</code> package, for 450k arrays, the relevant genomic regions to be analyzed are in files <code>450k_Gene_3_200.RDS</code> and <code>450k_InterGene_3_200.rds</code>. For EPIC arrays, the relevant genomic regions are in files <code>EPIC_Gene_3_200.rds</code> and <code>EPIC_InterGene_3_200.rds</code>. These additional data sets are available at <a href="https://github.com/TransBioInfoLab/coMethDMR_data" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_data</a>.</p>
<p>These files were created using the function <code>WriteCloseByAllRegions</code>, briefly, for genic regions, within each gene, we identified clusters of CpGs located closely (i.e. the maximum separation between any two consecutive probes is 200bp; <code>maxGap = 200</code>), and we required each cluster to have at least 3 CpGs (<code>minCpGs = 3</code>). For intergenic regions, we identified clusters CpGs similarly for each chromosome. To extract clusters of close-by CpGs from pre-defined genomic regions with different values of <code>maxGap</code> and <code>minCpGs</code>, the <code>WriteCloseByAllRegions</code> function can be used.</p>
<p>The pre-computed genomic regions can be accessed using the following commands. For geneic regions in 450k arrays,</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gene_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"450k_Gene_3_200.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>Here <code>gene_ls</code> is a list, with each item containing a character vector of CpGs IDs for a particular region in a gene.</p>
<p>Vignette # 2 illustrates how to leverage parallel computing via <code>BiocParallel</code> R package to make gene-based analysis fast.</p>
</div>
<div class="section level3">
<h3 id="when-there-are-co-variate-variables-in-dataset-to-consider">When there are co-variate variables in dataset to consider<a class="anchor" aria-label="anchor" href="#when-there-are-co-variate-variables-in-dataset-to-consider"></a>
</h3>
<p>Before identifying co-methylated clusters, we recommend removing uninteresting technical and biological effects, so that the resulting co-methylated clusters are only driven by the biological factors we are interested in. This can be accomplished using the <code>GetResiduals</code> function.</p>
<p>For example, the following script computes residuals from linear model <code>Mvalues ~ age.brain + sex + slide</code> (note that we use only a subset of the Chromosome 22 CpG islands for computing time consideration).</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Cgi_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"CpGislandsChr22_ex.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>
<span class="op">)</span>
<span class="va">betasChr22_df</span> <span class="op">&lt;-</span>
  <span class="va">betasChr22_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">Cgi_ls</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span>, <span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resid_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetResiduals.html">GetResiduals</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">betasChr22_df</span>,
  <span class="co"># converts to Mvalues for fitting linear model </span>
  betaToM <span class="op">=</span> <span class="cn">TRUE</span>, 
  pheno_df <span class="op">=</span> <span class="va">pheno_df</span>,
  covariates_char <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age.brain"</span>, <span class="st">"sex"</span>, <span class="st">"slide"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Phenotype data is not in the same order as methylation data. We used column Sample in phenotype data to put these two files in the same order.</span></code></pre>
</div>
<div class="section level3">
<h3 id="algorithm-for-identifying-co-methylated-regions">Algorithm for identifying co-methylated regions<a class="anchor" aria-label="anchor" href="#algorithm-for-identifying-co-methylated-regions"></a>
</h3>
<p>Within each genomic region, coMethDMR identifies contiguous and co-methylated CpGs sub-regions without using any outcome information. To select these co-methylated sub-regions, we use the <code>rdrop</code> statistic, which is the correlation between each CpG with the sum of methylation levels in all other CpGs. The default is <code>rDropThresh_num = 0.4</code>. We recommend this setting based on our simulation study. Note that higher <code>rDropThresh_num</code> values lead to fewer co-methylated regions.</p>
<p>Again, for illustration, we use CpG islands. For example, if we are interested in identifying co-methylated sub-region within the first genomic region in <code>Cgi_ls</code>:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Cgi_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"CpGislandsChr22_ex.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
    <span class="op">)</span>
<span class="op">)</span>

<span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">resid_mat</span>,
  betaToM <span class="op">=</span> <span class="cn">FALSE</span>,
  method <span class="op">=</span> <span class="st">"pearson"</span>,
  CpGs_ls <span class="op">=</span> <span class="va">Cgi_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>,
  output <span class="op">=</span> <span class="st">"CpGs"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMeth_ls</span></code></pre></div>
<pre><code><span class="co">## NULL</span></code></pre>
<p>The results (<code>NULL</code>) indicate there is no co-methylated sub-region within the first genomic region.</p>
<p>What about the other 19 regions? Next we look at a region (5th region in <code>Cgi_ls</code>) where there is a co-methylated sub-region:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">resid_mat</span>,
  betaToM <span class="op">=</span> <span class="cn">FALSE</span>,
  CpGs_ls <span class="op">=</span> <span class="va">Cgi_ls</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>, 
  output <span class="op">=</span> <span class="st">"CpGs"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMeth_ls</span></code></pre></div>
<pre><code><span class="co">## $`chr22:17565612-17565675`</span>
<span class="co">## [1] "cg21717745" "cg02866761" "cg05444620"</span></code></pre>
<p><code>coMeth_ls</code> is a list, where each item is a list of CpG probe IDs for a co-methylated sub-region.</p>
<p>If we want to see the detailed output of the coMethDMR algorithm, that is, how the co-methylated region was obtained, we can specify <code>output = "dataframe"</code>:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMethData_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">resid_mat</span>,
  betaToM <span class="op">=</span> <span class="cn">FALSE</span>,
  CpGs_ls <span class="op">=</span> <span class="va">Cgi_ls</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>, 
  output <span class="op">=</span> <span class="st">"dataframe"</span>
<span class="op">)</span> <span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMethData_df</span></code></pre></div>
<pre><code><span class="co">##                    Region        CpG   Chr  MAPINFO    r_drop keep</span>
<span class="co">## 1 chr22:17565612-17565904 cg21717745 chr22 17565612 0.7850186    1</span>
<span class="co">## 2 chr22:17565612-17565904 cg02866761 chr22 17565664 0.5547508    1</span>
<span class="co">## 3 chr22:17565612-17565904 cg05444620 chr22 17565675 0.6494019    1</span>
<span class="co">## 4 chr22:17565612-17565904 cg03304299 chr22 17565742 0.3282093    0</span>
<span class="co">##   keep_contiguous</span>
<span class="co">## 1               1</span>
<span class="co">## 2               1</span>
<span class="co">## 3               1</span>
<span class="co">## 4               0</span></code></pre>
<p><code>coMethData_df</code> provides the details on how the co-methylated region was obtained: Here <code>keep = 1</code> if <code>rDropThresh_num &gt; 0.4</code> (i.e. a co-methylated CpG), and <code>keep_contigous</code> indicates if the probe is in a contiguous co-methylated region. Note that only the last 3 CpGs constitutes the co-methylated cluster.</p>
</div>
<div class="section level3">
<h3 id="models-for-testing-genomic-regions-against-a-continuous-phenotype">Models for testing genomic regions against a continuous phenotype<a class="anchor" aria-label="anchor" href="#models-for-testing-genomic-regions-against-a-continuous-phenotype"></a>
</h3>
<p>To test association between a continuous phenotype and methylation values in a contiguous co-methylated region, two mixed models have been implemented in the function <code>lmmTestAllRegions</code>: a random coefficient mixed model (<code>modelType = "randCoef"</code>) and a simple linear mixed model (<code>modelType = "simple"</code>).</p>
<p>The random coefficient mixed model includes both a systematic component that models the mean for each group of CpGs, and a random component that models how each CpG varies with respect to the group mean (random probe effects). It also includes random sample effects that model correlations between multiple probes within the same sample.</p>
<p>More specifically, the random coefficient model is <code>methylation M value ~ contPheno_char + covariates_char + (1|Sample) + (contPheno_char|CpG).</code> The last term <code>(contPheno_char|CpG)</code> specifies both random intercepts and slopes for each CpG.</p>
<p>The simple linear mixed model includes all the terms in the random coefficient model except random probe effects.</p>
<p>The simple linear mixed model is</p>
<p><code>methylation M value ~ contPheno_char + covariates_char + (1|Sample)</code></p>
<p>To test one genomic region against the continuous phenotype <code>stage</code>, adjusting for <code>age.brain</code>:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/lmmTestAllRegions.html">lmmTestAllRegions</a></span><span class="op">(</span>
  betas <span class="op">=</span> <span class="va">betasChr22_df</span>,
  region_ls <span class="op">=</span> <span class="va">coMeth_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,
  <span class="va">pheno_df</span>,
  contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
  covariates_char <span class="op">=</span> <span class="st">"age.brain"</span>,
  modelType <span class="op">=</span> <span class="st">"randCoef"</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   chrom    start      end nCpGs   Estimate     StdErr     Stat   pValue</span>
<span class="co">## 1 chr22 17565612 17565675     3 0.03324127 0.02058149 1.615105 0.106288</span>
<span class="co">##        FDR</span>
<span class="co">## 1 0.106288</span></code></pre>
<p>If we don’t want to adjust for any covariate effect, we can set <code>covariates_char</code> to <code>NULL</code>.</p>
</div>
<div class="section level3">
<h3 id="analyzing-a-specific-gene">Analyzing a specific gene<a class="anchor" aria-label="anchor" href="#analyzing-a-specific-gene"></a>
</h3>
<p>Finally, we demonstrate <code>coMethDMR</code> analysis for a particular gene, for example the <code>ARFGAP3</code> gene.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span></code></pre></div>
<p>We assume that the user knows the set of probes corresponding to the gene of interest. If this is not the case, we provide two data sets which contain mappings from gene symbols to probe IDs for both 450k (“450k_CpGstoGene_min3CpGs.rds”) and EPIC (“EPIC_CpGstoGene_min3CpGs.rds”) arrays. These two data sets are available at: <a href="https://github.com/TransBioInfoLab/coMethDMR_data/tree/main/data" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_data/tree/main/data</a>.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># list probes for this gene</span>
<span class="va">ARFGAP3_CpGs_char</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
  <span class="st">"cg00079563"</span>, <span class="st">"cg01029450"</span>, <span class="st">"cg02351223"</span>, <span class="st">"cg04527868"</span>, <span class="st">"cg09861871"</span>,
  <span class="st">"cg26529516"</span>, <span class="st">"cg00539564"</span>, <span class="st">"cg05288033"</span>, <span class="st">"cg09367092"</span>, <span class="st">"cg10648908"</span>,
  <span class="st">"cg14570855"</span>, <span class="st">"cg15656623"</span>, <span class="st">"cg23778094"</span>, <span class="st">"cg27120833"</span>
<span class="op">)</span>

<span class="co"># list probes located closely on this gene</span>
<span class="va">gene3_200</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CloseBySingleRegion.html">CloseBySingleRegion</a></span><span class="op">(</span>
  CpGs_char <span class="op">=</span> <span class="va">ARFGAP3_CpGs_char</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  maxGap <span class="op">=</span> <span class="fl">200</span>,
  minCpGs <span class="op">=</span> <span class="fl">3</span>
<span class="op">)</span>
<span class="va">CpGsOrdered_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>
  <span class="va">gene3_200</span>,
  <span class="va">OrderCpGsByLocation</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  output <span class="op">=</span> <span class="st">"dataframe"</span>
<span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene3_200</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">CpGsOrdered_ls</span>, <span class="va">NameRegion</span><span class="op">)</span>

<span class="co"># List of regions</span>
<span class="va">gene3_200</span></code></pre></div>
<pre><code><span class="co">## $`chr22:43253145-43253208`</span>
<span class="co">## [1] "cg15656623" "cg10648908" "cg02351223"</span>
<span class="co">## </span>
<span class="co">## $`chr22:43253517-43253559`</span>
<span class="co">## [1] "cg09861871" "cg00079563" "cg26529516" "cg04527868" "cg01029450"</span>
<span class="co">## </span>
<span class="co">## $`chr22:43254043-43254168`</span>
<span class="co">## [1] "cg23778094" "cg00539564" "cg09367092"</span></code></pre>
<p>Now that we have a list of regions, we can find the co-methylated regions within the gene and test their statistical significance:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># co-methlyated region within the gene</span>
<span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">betasChr22_df</span>,
  betaToM <span class="op">=</span> <span class="cn">TRUE</span>,
  method <span class="op">=</span> <span class="st">"pearson"</span>,
  CpGs_ls <span class="op">=</span> <span class="va">gene3_200</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>,
  output <span class="op">=</span> <span class="st">"CpGs"</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span></code></pre>
<pre><code><span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">coMeth_ls</span> </code></pre></div>
<pre><code><span class="co">## $`chr22:43253521-43253559`</span>
<span class="co">## [1] "cg00079563" "cg26529516" "cg04527868" "cg01029450"</span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># test the co-methylated regions within the gene</span>
<span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmTestAllRegions.html">lmmTestAllRegions</a></span><span class="op">(</span>
  betas <span class="op">=</span> <span class="va">betasChr22_df</span>,
  region_ls <span class="op">=</span> <span class="va">coMeth_ls</span>,
  <span class="va">pheno_df</span>,
  contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
  covariates_char <span class="op">=</span> <span class="st">"age.brain"</span>,
  modelType <span class="op">=</span> <span class="st">"randCoef"</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>
  <span class="co"># generates a log file in the current directory</span>
  <span class="co"># outLogFile = paste0("lmmLog_", Sys.Date(), ".txt")</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-05-18</span>
<span class="co">## see ?sesameData and browseVignettes('sesameData') for documentation</span></code></pre>
<pre><code><span class="co">## Analyzing region chr22:43253521-43253559.</span></code></pre>
<pre><code><span class="co">## For future calls to this function, perhaps specify a log file.</span>
<span class="co">##       Set the file name of the log file with the outLogFile argument.</span></code></pre>
<p>Before we inspect the results, we will annotate them:</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/AnnotateResults.html">AnnotateResults</a></span><span class="op">(</span>lmmRes_df <span class="op">=</span> <span class="va">results</span>, arrayType <span class="op">=</span> <span class="st">"450k"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##   chrom    start      end nCpGs     Estimate     StdErr       Stat    pValue</span>
<span class="co">## 1 chr22 43253521 43253559     4 -0.006608311 0.02820896 -0.2342628 0.8147809</span>
<span class="co">##         FDR UCSC_RefGene_Group UCSC_RefGene_Accession UCSC_RefGene_Name</span>
<span class="co">## 1 0.8147809             TSS200 NM_001142293;NM_014570           ARFGAP3</span>
<span class="co">##   Relation_to_Island</span>
<span class="co">## 1             Island</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="frequently-asked-questions">Frequently Asked Questions<a class="anchor" aria-label="anchor" href="#frequently-asked-questions"></a>
</h2>
<ol style="list-style-type: decimal">
<li>What happens when mixed model fails to coverge (i.e. the warning “Model failed to converge with…” is resulted for a particular genomic region)?</li>
</ol>
<ul>
<li>In this case, the p-value for mixed model is set to 1. In our experiences with methylation datasets, genomic regions with strong signals typically converge. Convergence issues typically occurs when the amount of noise in data is high.</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>When fitting mixed models with <code>lmmTestAllRegions</code> function, What does the message “boundary (singular) fit” mean?</li>
</ol>
<ul>
<li>When mixed model is singular, at least one of the estimated variance components for intercepts or slopes random effects is 0, because there isn’t enough variabilities in data to estimate the random effects. In this case, mixed model reduces to a fixed effects model. However, as our simulation studies have shown, the p-values obtained for these regions are still valid.</li>
</ul>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>Lunnon K, Smith R, Hannon E, De Jager PL, Srivastava G, Volta M, Troakes C, Al-Sarraj S, Burrage J, Macdonald R, et al (2014) Methylomic profiling implicates cortical deregulation of ANK1 in Alzheimer’s disease. Nat Neurosci 17:1164-1170.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.1 (2021-08-10)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19042)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">## [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">## [3] LC_MONETARY=English_United States.1252</span>
<span class="co">## [4] LC_NUMERIC=C                          </span>
<span class="co">## [5] LC_TIME=English_United States.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] corrplot_0.92       sesameData_1.13.3   rmarkdown_2.11     </span>
<span class="co">##  [4] ExperimentHub_2.0.0 AnnotationHub_3.0.2 BiocFileCache_2.0.0</span>
<span class="co">##  [7] dbplyr_2.1.1        BiocGenerics_0.38.0 coMethDMR_0.99.17  </span>
<span class="co">## [10] BiocStyle_2.20.2   </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.4                                 </span>
<span class="co">##   [2] plyr_1.8.6                                        </span>
<span class="co">##   [3] splines_4.1.1                                     </span>
<span class="co">##   [4] BiocParallel_1.26.2                               </span>
<span class="co">##   [5] GenomeInfoDb_1.28.4                               </span>
<span class="co">##   [6] ggplot2_3.3.5                                     </span>
<span class="co">##   [7] digest_0.6.29                                     </span>
<span class="co">##   [8] foreach_1.5.2                                     </span>
<span class="co">##   [9] htmltools_0.5.2                                   </span>
<span class="co">##  [10] lmerTest_3.1-3                                    </span>
<span class="co">##  [11] fansi_1.0.2                                       </span>
<span class="co">##  [12] magrittr_2.0.2                                    </span>
<span class="co">##  [13] memoise_2.0.1                                     </span>
<span class="co">##  [14] tzdb_0.2.0                                        </span>
<span class="co">##  [15] limma_3.48.3                                      </span>
<span class="co">##  [16] readr_2.1.2                                       </span>
<span class="co">##  [17] annotate_1.70.0                                   </span>
<span class="co">##  [18] Biostrings_2.60.2                                 </span>
<span class="co">##  [19] matrixStats_0.61.0                                </span>
<span class="co">##  [20] askpass_1.1                                       </span>
<span class="co">##  [21] siggenes_1.66.0                                   </span>
<span class="co">##  [22] pkgdown_2.0.2                                     </span>
<span class="co">##  [23] prettyunits_1.1.1                                 </span>
<span class="co">##  [24] colorspace_2.0-3                                  </span>
<span class="co">##  [25] blob_1.2.2                                        </span>
<span class="co">##  [26] rappdirs_0.3.3                                    </span>
<span class="co">##  [27] textshaping_0.3.6                                 </span>
<span class="co">##  [28] xfun_0.29                                         </span>
<span class="co">##  [29] dplyr_1.0.8                                       </span>
<span class="co">##  [30] crayon_1.5.0                                      </span>
<span class="co">##  [31] RCurl_1.98-1.6                                    </span>
<span class="co">##  [32] jsonlite_1.7.3                                    </span>
<span class="co">##  [33] genefilter_1.74.1                                 </span>
<span class="co">##  [34] lme4_1.1-28                                       </span>
<span class="co">##  [35] GEOquery_2.60.0                                   </span>
<span class="co">##  [36] survival_3.2-13                                   </span>
<span class="co">##  [37] iterators_1.0.14                                  </span>
<span class="co">##  [38] glue_1.6.1                                        </span>
<span class="co">##  [39] gtable_0.3.0                                      </span>
<span class="co">##  [40] zlibbioc_1.38.0                                   </span>
<span class="co">##  [41] XVector_0.32.0                                    </span>
<span class="co">##  [42] DelayedArray_0.18.0                               </span>
<span class="co">##  [43] Rhdf5lib_1.14.2                                   </span>
<span class="co">##  [44] HDF5Array_1.20.0                                  </span>
<span class="co">##  [45] scales_1.1.1                                      </span>
<span class="co">##  [46] DBI_1.1.2                                         </span>
<span class="co">##  [47] rngtools_1.5.2                                    </span>
<span class="co">##  [48] Rcpp_1.0.8                                        </span>
<span class="co">##  [49] xtable_1.8-4                                      </span>
<span class="co">##  [50] progress_1.2.2                                    </span>
<span class="co">##  [51] bumphunter_1.34.0                                 </span>
<span class="co">##  [52] mclust_5.4.9                                      </span>
<span class="co">##  [53] bit_4.0.4                                         </span>
<span class="co">##  [54] preprocessCore_1.54.0                             </span>
<span class="co">##  [55] stats4_4.1.1                                      </span>
<span class="co">##  [56] httr_1.4.2                                        </span>
<span class="co">##  [57] RColorBrewer_1.1-2                                </span>
<span class="co">##  [58] ellipsis_0.3.2                                    </span>
<span class="co">##  [59] reshape_0.8.8                                     </span>
<span class="co">##  [60] pkgconfig_2.0.3                                   </span>
<span class="co">##  [61] XML_3.99-0.8                                      </span>
<span class="co">##  [62] sass_0.4.0                                        </span>
<span class="co">##  [63] locfit_1.5-9.4                                    </span>
<span class="co">##  [64] utf8_1.2.2                                        </span>
<span class="co">##  [65] tidyselect_1.1.2                                  </span>
<span class="co">##  [66] rlang_1.0.1                                       </span>
<span class="co">##  [67] later_1.3.0                                       </span>
<span class="co">##  [68] AnnotationDbi_1.54.1                              </span>
<span class="co">##  [69] munsell_0.5.0                                     </span>
<span class="co">##  [70] BiocVersion_3.13.1                                </span>
<span class="co">##  [71] tools_4.1.1                                       </span>
<span class="co">##  [72] cachem_1.0.6                                      </span>
<span class="co">##  [73] cli_3.2.0                                         </span>
<span class="co">##  [74] generics_0.1.2                                    </span>
<span class="co">##  [75] RSQLite_2.2.10                                    </span>
<span class="co">##  [76] evaluate_0.15                                     </span>
<span class="co">##  [77] stringr_1.4.0                                     </span>
<span class="co">##  [78] fastmap_1.1.0                                     </span>
<span class="co">##  [79] yaml_2.3.5                                        </span>
<span class="co">##  [80] ragg_1.2.2                                        </span>
<span class="co">##  [81] knitr_1.37                                        </span>
<span class="co">##  [82] bit64_4.0.5                                       </span>
<span class="co">##  [83] fs_1.5.2                                          </span>
<span class="co">##  [84] beanplot_1.2                                      </span>
<span class="co">##  [85] scrime_1.3.5                                      </span>
<span class="co">##  [86] purrr_0.3.4                                       </span>
<span class="co">##  [87] KEGGREST_1.32.0                                   </span>
<span class="co">##  [88] sparseMatrixStats_1.4.2                           </span>
<span class="co">##  [89] nlme_3.1-152                                      </span>
<span class="co">##  [90] doRNG_1.8.2                                       </span>
<span class="co">##  [91] mime_0.12                                         </span>
<span class="co">##  [92] nor1mix_1.3-0                                     </span>
<span class="co">##  [93] xml2_1.3.3                                        </span>
<span class="co">##  [94] biomaRt_2.48.3                                    </span>
<span class="co">##  [95] compiler_4.1.1                                    </span>
<span class="co">##  [96] rstudioapi_0.13                                   </span>
<span class="co">##  [97] filelock_1.0.2                                    </span>
<span class="co">##  [98] curl_4.3.2                                        </span>
<span class="co">##  [99] png_0.1-7                                         </span>
<span class="co">## [100] interactiveDisplayBase_1.30.0                     </span>
<span class="co">## [101] tibble_3.1.6                                      </span>
<span class="co">## [102] bslib_0.3.1                                       </span>
<span class="co">## [103] stringi_1.7.6                                     </span>
<span class="co">## [104] highr_0.9                                         </span>
<span class="co">## [105] GenomicFeatures_1.44.2                            </span>
<span class="co">## [106] desc_1.4.0                                        </span>
<span class="co">## [107] minfi_1.38.0                                      </span>
<span class="co">## [108] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0</span>
<span class="co">## [109] lattice_0.20-44                                   </span>
<span class="co">## [110] Matrix_1.3-4                                      </span>
<span class="co">## [111] nloptr_2.0.0                                      </span>
<span class="co">## [112] multtest_2.48.0                                   </span>
<span class="co">## [113] vctrs_0.3.8                                       </span>
<span class="co">## [114] rhdf5filters_1.4.0                                </span>
<span class="co">## [115] pillar_1.7.0                                      </span>
<span class="co">## [116] lifecycle_1.0.1                                   </span>
<span class="co">## [117] BiocManager_1.30.16                               </span>
<span class="co">## [118] jquerylib_0.1.4                                   </span>
<span class="co">## [119] data.table_1.14.2                                 </span>
<span class="co">## [120] bitops_1.0-7                                      </span>
<span class="co">## [121] httpuv_1.6.5                                      </span>
<span class="co">## [122] rtracklayer_1.52.1                                </span>
<span class="co">## [123] GenomicRanges_1.44.0                              </span>
<span class="co">## [124] R6_2.5.1                                          </span>
<span class="co">## [125] BiocIO_1.2.0                                      </span>
<span class="co">## [126] bookdown_0.24                                     </span>
<span class="co">## [127] promises_1.2.0.1                                  </span>
<span class="co">## [128] IRanges_2.26.0                                    </span>
<span class="co">## [129] codetools_0.2-18                                  </span>
<span class="co">## [130] boot_1.3-28                                       </span>
<span class="co">## [131] MASS_7.3-54                                       </span>
<span class="co">## [132] assertthat_0.2.1                                  </span>
<span class="co">## [133] rhdf5_2.36.0                                      </span>
<span class="co">## [134] SummarizedExperiment_1.22.0                       </span>
<span class="co">## [135] openssl_1.4.6                                     </span>
<span class="co">## [136] rprojroot_2.0.2                                   </span>
<span class="co">## [137] rjson_0.2.21                                      </span>
<span class="co">## [138] withr_2.4.3                                       </span>
<span class="co">## [139] GenomicAlignments_1.28.0                          </span>
<span class="co">## [140] Rsamtools_2.8.0                                   </span>
<span class="co">## [141] S4Vectors_0.30.2                                  </span>
<span class="co">## [142] GenomeInfoDbData_1.2.6                            </span>
<span class="co">## [143] hms_1.1.1                                         </span>
<span class="co">## [144] quadprog_1.5-8                                    </span>
<span class="co">## [145] grid_4.1.1                                        </span>
<span class="co">## [146] tidyr_1.2.0                                       </span>
<span class="co">## [147] base64_2.0                                        </span>
<span class="co">## [148] minqa_1.2.4                                       </span>
<span class="co">## [149] DelayedMatrixStats_1.14.3                         </span>
<span class="co">## [150] illuminaio_0.34.0                                 </span>
<span class="co">## [151] MatrixGenerics_1.4.3                              </span>
<span class="co">## [152] numDeriv_2016.8-1.1                               </span>
<span class="co">## [153] Biobase_2.52.0                                    </span>
<span class="co">## [154] shiny_1.7.1                                       </span>
<span class="co">## [155] restfulr_0.0.13</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Fernanda Veitzman, Lissette Gomez, Tiago Silva, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d" class="external-link">Lily Wang</a>, <a href="https://github.com/gabrielodom" class="external-link">Gabriel Odom</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
