<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genome-wide methylation analysis using coMethDMR via parallel computing • coMethDMR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Genome-wide methylation analysis using coMethDMR via parallel computing">
<meta property="og:description" content="coMethDMR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coMethDMR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vin1_Introduction_to_coMethDMR_geneBasedPipeline.html">Introduction to coMethDMR</a>
    </li>
    <li>
      <a href="../articles/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.html">Genome-wide methylation analysis using coMethDMR via parallel computing</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/TransBioInfoLab/coMethDMR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Genome-wide methylation analysis using coMethDMR via parallel computing</h1>
                        <h4 data-toc-skip class="author">Gabriel J. Odom, Lissette Gomez, Tiago Chedraoui Silva, Lanyu Zhang and Lily Wang</h4>
            
            <h4 data-toc-skip class="date">March 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/TransBioInfoLab/coMethDMR/blob/HEAD/vignettes/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.Rmd" class="external-link"><code>vignettes/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.Rmd</code></a></small>
      <div class="hidden name"><code>vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In the previous vignette “Introduction to <strong>coMethDMR</strong>”, we discussed components of the <strong>coMethDMR</strong> pipeline and presented example R scripts for running an analysis with <strong>coMethDMR</strong> serially. However, because identifying co-methylated clusters and fitting mixed effects models on large numbers of genomic regions can be computationally expensive, we illustrate implementation of parallel computing for <strong>coMethDMR</strong> via the <strong>BiocParallel</strong> R package in this vignette.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The latest version can be installed by:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"coMethDMR"</span><span class="op">)</span></code></pre></div>
<p>First, we load these two packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gabrielodom.github.io/coMethDMR/" class="external-link">coMethDMR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>In Section 2, we give a brief re-introduction to the example data. In Section 3, we present example scripts for analyzing a single type (e.g. genic regions) of genomic regions using parallel computing. In Section 4, we present example scripts for analyzing genic and intergenic regions on the Illumina arrays using parallel computing.</p>
</div>
</div>
<div class="section level2">
<h2 id="example-dataset">Example Dataset<a class="anchor" aria-label="anchor" href="#example-dataset"></a>
</h2>
<p>For illustration, we use a subset of prefrontal cortex methylation data (GEO GSE59685) described in Lunnon et al. (2014). This example dataset includes beta values for 8552 CpGs on chromosome 22 for a random selection of 20 subjects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span></code></pre></div>
<p>We assume quality control and normalization of the methylation dataset have been performed by R packages such as <strong>minfi</strong> or <strong>RnBeads</strong>. Further, we assume that probes and samples with excessive missingness have been removed (this task can be completed by subsetting the original methylation data set with the probes and samples identified by the <code><a href="../reference/MarkMissing.html">MarkMissing()</a></code> function).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Purge any probes or samples with excessive missing values</span>
<span class="va">markCells_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MarkMissing.html">MarkMissing</a></span><span class="op">(</span>dnaM_df <span class="op">=</span> <span class="va">betasChr22_df</span><span class="op">)</span>
<span class="va">betasChr22_df</span> <span class="op">&lt;-</span> <span class="va">betasChr22_df</span><span class="op">[</span>
  <span class="va">markCells_ls</span><span class="op">$</span><span class="va">keepProbes</span>, 
  <span class="va">markCells_ls</span><span class="op">$</span><span class="va">keepSamples</span>
<span class="op">]</span>

<span class="co"># Inspect</span>
<span class="va">betasChr22_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##            GSM1443279 GSM1443663 GSM1443434 GSM1443547 GSM1443577</span>
<span class="co">## cg00004192  0.9249942  0.8463296  0.8700718  0.9058205  0.9090382</span>
<span class="co">## cg00004775  0.6523025  0.6247554  0.7573476  0.6590817  0.6726261</span>
<span class="co">## cg00012194  0.8676339  0.8679048  0.8484754  0.8754985  0.8484458</span>
<span class="co">## cg00013618  0.9466056  0.9475467  0.9566493  0.9588431  0.9419563</span>
<span class="co">## cg00014104  0.3932388  0.5525716  0.4075900  0.3997278  0.3216956</span></code></pre>
<p>The corresponding phenotype dataset included variables <code>stage</code> (Braak AD stage), <code>subject.id</code>, <code>slide</code> (batch effect), <code>sex</code>, <code>Sample</code> and <code>age.brain</code> (age of the brain donor).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pheno_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pheno_df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    stage subject.id         sex     Sample age.brain       slide</span>
<span class="co">## 3      0          1 Sex: FEMALE GSM1443251        82  6042316048</span>
<span class="co">## 8      2          2 Sex: FEMALE GSM1443256        82  6042316066</span>
<span class="co">## 10    NA          3   Sex: MALE GSM1443258        89  6042316066</span>
<span class="co">## 15     1          4 Sex: FEMALE GSM1443263        81  7786923107</span>
<span class="co">## 21     2          5 Sex: FEMALE GSM1443269        92  6042316121</span>
<span class="co">## 22     1          6   Sex: MALE GSM1443270        78  6042316099</span></code></pre>
<p>Note that only samples with both methylation data and non-missing phenotype data are analyzed by <strong>coMethDMR</strong>. So in this example, the sample with <code>subject.id = 3</code> which lacks AD stage information will be excluded from analysis. Please also note the phenotype file needs to have a variable called “Sample” that will be used by <strong>coMethDMR</strong> to link to the methylation dataset.</p>
</div>
<div class="section level2">
<h2 id="analyzing-one-type-of-genomic-region-via-biocparallel">Analyzing One Type of Genomic Region via BiocParallel<a class="anchor" aria-label="anchor" href="#analyzing-one-type-of-genomic-region-via-biocparallel"></a>
</h2>
<p>As mentioned previously in “Introduction to <strong>coMethDMR</strong>”, there are several steps in the <strong>coMethDMR</strong> pipeline:</p>
<ol style="list-style-type: decimal">
<li>Obtain CpGs located closely in pre-defined genomic regions,</li>
<li>Identify co-methylated regions, and</li>
<li>Test co-methylated regions against the outcome variable (AD stage).</li>
</ol>
<p>Suppose we are interested in analyzing genic regions on the 450k arrays. In the following, <code>closeByGeneAll_ls</code> is a list, where each item includes at least 3 CpGs located closely (max distance between 2 CpGs is 200bp by default).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">closeByGeneAll_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"450k_Gene_3_200.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>We can inspect the first region in the list via:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">closeByGeneAll_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## $`chr4:91760141-91760229`</span>
<span class="co">## [1] "cg01583657" "cg25325192" "cg06044899"</span></code></pre>
<p>Next, for demonstration, we select regions on chromosome 22.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">indx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"chr22:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">closeByGeneAll_ls</span><span class="op">)</span><span class="op">)</span>

<span class="va">closeByGene_ls</span> <span class="op">&lt;-</span> <span class="va">closeByGeneAll_ls</span><span class="op">[</span><span class="va">indx</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">closeByGeneAll_ls</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">closeByGene_ls</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 676</span></code></pre>
<p>There are 676 genic regions to be tested for chromosome 22. As an example, this vignette will analyze the first 10 regions, which contain the following CpGs:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">closeByGene_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## $`chr22:30662799-30663041`</span>
<span class="co">## [1] "cg19018155" "cg10467217" "cg24727122" "cg02597698" "cg25666403"</span>
<span class="co">## [6] "cg02026204" "cg05539509" "cg07498879"</span>
<span class="co">## </span>
<span class="co">## $`chr22:18632404-18633123`</span>
<span class="co">##  [1] "cg27281093" "cg18699242" "cg01963623" "cg02169692" "cg00077299"</span>
<span class="co">##  [6] "cg07964128" "cg02518889" "cg01246421" "cg08205655" "cg10685559"</span>
<span class="co">## </span>
<span class="co">## $`chr22:43253145-43253208`</span>
<span class="co">## [1] "cg15656623" "cg10648908" "cg02351223"</span>
<span class="co">## </span>
<span class="co">## $`chr22:43253517-43253559`</span>
<span class="co">## [1] "cg09861871" "cg00079563" "cg26529516" "cg04527868" "cg01029450"</span>
<span class="co">## </span>
<span class="co">## $`chr22:43254043-43254168`</span>
<span class="co">## [1] "cg23778094" "cg00539564" "cg09367092"</span>
<span class="co">## </span>
<span class="co">## $`chr22:30901532-30901886`</span>
<span class="co">## [1] "cg03214697" "cg00093544" "cg02923264" "cg03014622" "cg19964641"</span>
<span class="co">## [6] "cg15829116" "cg07217063"</span>
<span class="co">## </span>
<span class="co">## $`chr22:39883190-39883347`</span>
<span class="co">## [1] "cg00550928" "cg00101350" "cg26692811"</span>
<span class="co">## </span>
<span class="co">## $`chr22:30685584-30685926`</span>
<span class="co">## [1] "cg00101453" "cg23835894" "cg22088670" "cg03652890" "cg24843389"</span>
<span class="co">## [6] "cg15272908" "cg20395643" "cg10998744"</span>
<span class="co">## </span>
<span class="co">## $`chr22:45064086-45064238`</span>
<span class="co">## [1] "cg10031995" "cg21293611" "cg11450750"</span>
<span class="co">## </span>
<span class="co">## $`chr22:45072455-45072724`</span>
<span class="co">## [1] "cg24019054" "cg20339720" "cg16166559" "cg06849477"</span></code></pre>
<p>In order to make these examples as simple as possible, we will remove the CpGs from our data set which are not included in these 10 regions. In practice, you would keep the full data set.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keepCpGs_char</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">closeByGene_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">betasChr22small_df</span> <span class="op">&lt;-</span> 
  <span class="va">betasChr22_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">keepCpGs_char</span>, <span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">betasChr22small_df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 54 20</span></code></pre>
<div class="section level3">
<h3 id="compute-residuals">Compute residuals<a class="anchor" aria-label="anchor" href="#compute-residuals"></a>
</h3>
<p>This step removes uninteresting technical and biological effects using the <code>GetResiduals</code> function, so that the resulting co-methylated clusters are only driven by the biological factors we are interested in. Note that there may be an error with the parallel back-end (<code>1 parallel job did not deliver a result</code>); this has occurred rarely for us (and not consistently enough for us to create a reproducible example), but it goes away when we execute the code a second time. We then recommend that you test this with a smaller subset of your data first. If you are able to consistently replicate a parallel error, please submit a bug report: <a href="https://github.com/TransBioInfoLab/coMethDMR/issues" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR/issues</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resid_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetResiduals.html">GetResiduals</a></span><span class="op">(</span>
  dnam <span class="op">=</span> <span class="va">betasChr22small_df</span>,
  <span class="co"># converts to Mvalues for fitting linear model</span>
  betaToM <span class="op">=</span> <span class="cn">TRUE</span>,  
  pheno_df <span class="op">=</span> <span class="va">pheno_df</span>,
  <span class="co"># Features in pheno_df used as covariates</span>
  covariates_char <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age.brain"</span>, <span class="st">"sex"</span>, <span class="st">"slide"</span><span class="op">)</span>,
  nCores_int <span class="op">=</span> <span class="fl">2</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Phenotype data is not in the same order as methylation data. We used column Sample in phenotype data to put these two files in the same order.</span></code></pre>
</div>
<div class="section level3">
<h3 id="finding-co-methylated-regions">Finding co-methylated regions<a class="anchor" aria-label="anchor" href="#finding-co-methylated-regions"></a>
</h3>
<p>The cluster computing with the <strong>BiocParallel</strong> package involves simply changing the default value of one argument: <code>nCores_int</code>. This argument enables you to perform parallel computing when you set the number of cores to an integer value greater than 1. <em>If you do not know how many cores your machine has, use the <code>detectCores()</code> function from the <code>parallel</code> package. Note that, if you have many applications open on your computer, you should not use all of the cores available.</em></p>
<p>Once we know how many cores we have available, we execute the <code><a href="../reference/CoMethAllRegions.html">CoMethAllRegions()</a></code> function using each worker in the cluster, to find co-methylated clusters in the genic regions. As an example, we only analyze the first 10 CpG regions (<code>CpGs_ls</code>).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span>
  <span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CoMethAllRegions.html">CoMethAllRegions</a></span><span class="op">(</span>
    dnam <span class="op">=</span> <span class="va">resid_mat</span>,
    betaToM <span class="op">=</span> <span class="cn">FALSE</span>,
    method <span class="op">=</span> <span class="st">"spearman"</span>,
    arrayType <span class="op">=</span> <span class="st">"450k"</span>,
    CpGs_ls <span class="op">=</span> <span class="va">closeByGene_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,
    nCores_int <span class="op">=</span> <span class="fl">2</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code><span class="co">##    user  system elapsed </span>
<span class="co">##   10.75    2.40   57.46</span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Windows: NA</span>
<span class="co"># Mac: ~14 seconds for 2 cores</span></code></pre></div>
<p>The object <code>coMeth_ls</code> is a list, with each item containing the list of CpGs within an identified co-methylated region.</p>
</div>
<div class="section level3">
<h3 id="testing-the-co-methylated-regions">Testing the co-methylated regions<a class="anchor" aria-label="anchor" href="#testing-the-co-methylated-regions"></a>
</h3>
<p>Next we test these co-methylated regions against continuous phenotype <code>stage</code>, adjusting for covariates <code>age</code> and <code>sex</code>, by executing the <code><a href="../reference/lmmTestAllRegions.html">lmmTestAllRegions()</a></code> function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lmmTestAllRegions.html">lmmTestAllRegions</a></span><span class="op">(</span>
  betas <span class="op">=</span> <span class="va">betasChr22small_df</span>,
  region_ls <span class="op">=</span> <span class="va">coMeth_ls</span>,
  pheno_df <span class="op">=</span> <span class="va">pheno_df</span>,
  contPheno_char <span class="op">=</span> <span class="st">"stage"</span>,
  covariates_char <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age.brain"</span>, <span class="st">"sex"</span><span class="op">)</span>,
  modelType <span class="op">=</span> <span class="st">"randCoef"</span>,
  arrayType <span class="op">=</span> <span class="st">"450k"</span>,
  nCores_int <span class="op">=</span> <span class="fl">2</span>
  <span class="co"># outLogFile = "res_lmm_log.txt"</span>
<span class="op">)</span></code></pre></div>
<p>Model fit messages and diagnostics for each region will be saved to the log file specified with the <code>outLogFile</code> argument. For a single region, this will return a one row of model fit statistics similar to the following:</p>
<table class="table">
<thead><tr class="header">
<th>chrom</th>
<th>start</th>
<th>end</th>
<th>nCpGs</th>
<th>Estimate</th>
<th>StdErr</th>
<th>Stat</th>
<th>pValue</th>
</tr></thead>
<tbody><tr class="odd">
<td>chr22</td>
<td>24823455</td>
<td>24823519</td>
<td>4</td>
<td>-0.0702</td>
<td>0.0290</td>
<td>-2.4184</td>
<td>0.0155</td>
</tr></tbody>
</table>
<p>Finally, we can annotate these results:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/AnnotateResults.html">AnnotateResults</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Setting options('download.file.method.GEOquery'='auto')</span></code></pre>
<pre><code><span class="co">## Setting options('GEOquery.inmemory.gpl'=FALSE)</span></code></pre>
<pre><code><span class="co">##   chrom    start      end nCpGs     Estimate     StdErr       Stat      pValue</span>
<span class="co">## 1 chr22 30662799 30663041     8 -0.059977631 0.01878112 -3.1935073 0.001405558</span>
<span class="co">## 2 chr22 43253521 43253559     4 -0.007602187 0.02873573 -0.2645552 0.791352119</span>
<span class="co">## 3 chr22 30901532 30901886     7  0.018015623 0.01555681  1.1580537 0.246842103</span>
<span class="co">## 4 chr22 39883190 39883347     3  0.069276861 0.02864166  2.4187449 0.015574157</span>
<span class="co">##           FDR           UCSC_RefGene_Group UCSC_RefGene_Accession</span>
<span class="co">## 1 0.005622231 1stExon;5'UTR;TSS1500;TSS200              NM_020530</span>
<span class="co">## 2 0.791352119                       TSS200 NM_001142293;NM_014570</span>
<span class="co">## 3 0.329122804    1stExon;5'UTR;Body;TSS200 NM_001161368;NM_174977</span>
<span class="co">## 4 0.031148314         1stExon;5'UTR;TSS200 NM_001098270;NM_002409</span>
<span class="co">##   UCSC_RefGene_Name Relation_to_Island</span>
<span class="co">## 1               OSM            OpenSea</span>
<span class="co">## 2           ARFGAP3             Island</span>
<span class="co">## 3           SEC14L4     Island;S_Shore</span>
<span class="co">## 4             MGAT3            N_Shore</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="comethdmr-analysis-pipeline-for-450k-methylation-arrays-datasets-via-biocparallel">coMethDMR Analysis Pipeline for 450k Methylation Arrays Datasets via BiocParallel<a class="anchor" aria-label="anchor" href="#comethdmr-analysis-pipeline-for-450k-methylation-arrays-datasets-via-biocparallel"></a>
</h2>
<p>In this section, we provide example scripts for testing genic and intergenic regions using <strong>coMethDMR</strong> and <strong>BiocParallel</strong> R packages.</p>
<p>First, we read in clusters of CpGs located closely on the array, in genic and intergenic regions, then combine them into a single list.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">closeByGene_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"450k_Gene_3_200.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>
<span class="op">)</span>

<span class="va">closeByInterGene_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
    <span class="st">"extdata"</span>,
    <span class="st">"450k_InterGene_3_200.rds"</span>,
    package <span class="op">=</span> <span class="st">'coMethDMR'</span>,
    mustWork <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>
<span class="op">)</span>

<span class="co"># put them together in one list</span>
<span class="va">closeBy_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">closeByInterGene_ls</span>, <span class="va">closeByGene_ls</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">closeBy_ls</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 42536</span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">closeBy_ls</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## $`chr1:569427-569687`</span>
<span class="co">## [1] "cg08858441" "cg03348902" "cg01070250"</span>
<span class="co">## </span>
<span class="co">## $`chr1:713921-714177`</span>
<span class="co">## [1] "cg10037654" "cg14057946" "cg11422233" "cg16047670" "cg18263059"</span>
<span class="co">## [6] "cg01014490"</span>
<span class="co">## </span>
<span class="co">## $`chr1:834183-834356`</span>
<span class="co">## [1] "cg13938959" "cg12445832" "cg23999112"</span></code></pre>
<p>With this list of regions and their probe IDs, repeat the analysis code in the previous section.</p>
</div>
<div class="section level2">
<h2 id="a-comment-on-using-epic-methylation-arrays-datasets">A Comment on Using EPIC Methylation Arrays Datasets<a class="anchor" aria-label="anchor" href="#a-comment-on-using-epic-methylation-arrays-datasets"></a>
</h2>
<p>The analysis for EPIC methylation arrays would be the same as those for 450k arrays, except by testing genomic regions in files “EPIC_Gene_3_200.rds” and “EPIC_InterGene_3_200.RDS”, instead of “450k_Gene_3_200.rds” and “450k_InterGene_3_200.rds”. These data sets are available at <a href="https://github.com/TransBioInfoLab/coMethDMR_data/tree/main/data" class="external-link uri">https://github.com/TransBioInfoLab/coMethDMR_data/tree/main/data</a>.</p>
</div>
<div class="section level2">
<h2 id="additional-comments-on-computational-time-and-resources">Additional Comments on Computational Time and Resources<a class="anchor" aria-label="anchor" href="#additional-comments-on-computational-time-and-resources"></a>
</h2>
<p>In this vignette, we have analyzed a small subset of a real EWAS dataset (i.e. only chromosome 22 data on 20 subjects). To give users a more realistic estimate of time for analyzing real EWAS datasets, we also measured time used for analyzing the entire Lunnon et al. (2014) dataset with 110 samples on all chromosomes. These computation times measured on a Dell Precision 5810 with 64Gb of RAM, an Intel Xeon E5-2640 CPU at 2.40Ghz, and using up to 20 cores. More specifically, in Section 4, the entire <strong>coMethDMR</strong> workflow for took 103 minutes with 6 cores and used a maximum of 24Gb of RAM (for the <code><a href="../reference/CoMethAllRegions.html">CoMethAllRegions()</a></code> function). We’re currently working improving the speed and reducing the size of <strong>coMethDMR</strong>, so please check back soon for updates.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.1 (2021-08-10)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19042)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">## [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">## [3] LC_MONETARY=English_United States.1252</span>
<span class="co">## [4] LC_NUMERIC=C                          </span>
<span class="co">## [5] LC_TIME=English_United States.1252    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] sesameData_1.13.3   rmarkdown_2.11      ExperimentHub_2.0.0</span>
<span class="co">##  [4] AnnotationHub_3.0.2 BiocFileCache_2.0.0 dbplyr_2.1.1       </span>
<span class="co">##  [7] BiocGenerics_0.38.0 BiocParallel_1.26.2 coMethDMR_0.99.18  </span>
<span class="co">## [10] BiocStyle_2.20.2   </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] snow_0.4-4                                        </span>
<span class="co">##   [2] systemfonts_1.0.4                                 </span>
<span class="co">##   [3] plyr_1.8.6                                        </span>
<span class="co">##   [4] splines_4.1.1                                     </span>
<span class="co">##   [5] GenomeInfoDb_1.28.4                               </span>
<span class="co">##   [6] ggplot2_3.3.5                                     </span>
<span class="co">##   [7] digest_0.6.29                                     </span>
<span class="co">##   [8] foreach_1.5.2                                     </span>
<span class="co">##   [9] htmltools_0.5.2                                   </span>
<span class="co">##  [10] lmerTest_3.1-3                                    </span>
<span class="co">##  [11] fansi_1.0.2                                       </span>
<span class="co">##  [12] magrittr_2.0.2                                    </span>
<span class="co">##  [13] memoise_2.0.1                                     </span>
<span class="co">##  [14] tzdb_0.2.0                                        </span>
<span class="co">##  [15] limma_3.48.3                                      </span>
<span class="co">##  [16] readr_2.1.2                                       </span>
<span class="co">##  [17] annotate_1.70.0                                   </span>
<span class="co">##  [18] Biostrings_2.60.2                                 </span>
<span class="co">##  [19] matrixStats_0.61.0                                </span>
<span class="co">##  [20] askpass_1.1                                       </span>
<span class="co">##  [21] siggenes_1.66.0                                   </span>
<span class="co">##  [22] pkgdown_2.0.2                                     </span>
<span class="co">##  [23] prettyunits_1.1.1                                 </span>
<span class="co">##  [24] colorspace_2.0-3                                  </span>
<span class="co">##  [25] blob_1.2.2                                        </span>
<span class="co">##  [26] rappdirs_0.3.3                                    </span>
<span class="co">##  [27] textshaping_0.3.6                                 </span>
<span class="co">##  [28] xfun_0.29                                         </span>
<span class="co">##  [29] dplyr_1.0.8                                       </span>
<span class="co">##  [30] crayon_1.5.0                                      </span>
<span class="co">##  [31] RCurl_1.98-1.6                                    </span>
<span class="co">##  [32] jsonlite_1.7.3                                    </span>
<span class="co">##  [33] genefilter_1.74.1                                 </span>
<span class="co">##  [34] lme4_1.1-28                                       </span>
<span class="co">##  [35] GEOquery_2.60.0                                   </span>
<span class="co">##  [36] survival_3.2-13                                   </span>
<span class="co">##  [37] iterators_1.0.14                                  </span>
<span class="co">##  [38] glue_1.6.1                                        </span>
<span class="co">##  [39] gtable_0.3.0                                      </span>
<span class="co">##  [40] zlibbioc_1.38.0                                   </span>
<span class="co">##  [41] XVector_0.32.0                                    </span>
<span class="co">##  [42] DelayedArray_0.18.0                               </span>
<span class="co">##  [43] Rhdf5lib_1.14.2                                   </span>
<span class="co">##  [44] HDF5Array_1.20.0                                  </span>
<span class="co">##  [45] scales_1.1.1                                      </span>
<span class="co">##  [46] DBI_1.1.2                                         </span>
<span class="co">##  [47] rngtools_1.5.2                                    </span>
<span class="co">##  [48] Rcpp_1.0.8                                        </span>
<span class="co">##  [49] xtable_1.8-4                                      </span>
<span class="co">##  [50] progress_1.2.2                                    </span>
<span class="co">##  [51] bumphunter_1.34.0                                 </span>
<span class="co">##  [52] mclust_5.4.9                                      </span>
<span class="co">##  [53] bit_4.0.4                                         </span>
<span class="co">##  [54] preprocessCore_1.54.0                             </span>
<span class="co">##  [55] stats4_4.1.1                                      </span>
<span class="co">##  [56] httr_1.4.2                                        </span>
<span class="co">##  [57] RColorBrewer_1.1-2                                </span>
<span class="co">##  [58] ellipsis_0.3.2                                    </span>
<span class="co">##  [59] reshape_0.8.8                                     </span>
<span class="co">##  [60] pkgconfig_2.0.3                                   </span>
<span class="co">##  [61] XML_3.99-0.8                                      </span>
<span class="co">##  [62] sass_0.4.0                                        </span>
<span class="co">##  [63] locfit_1.5-9.4                                    </span>
<span class="co">##  [64] utf8_1.2.2                                        </span>
<span class="co">##  [65] tidyselect_1.1.2                                  </span>
<span class="co">##  [66] rlang_1.0.1                                       </span>
<span class="co">##  [67] later_1.3.0                                       </span>
<span class="co">##  [68] AnnotationDbi_1.54.1                              </span>
<span class="co">##  [69] munsell_0.5.0                                     </span>
<span class="co">##  [70] BiocVersion_3.13.1                                </span>
<span class="co">##  [71] tools_4.1.1                                       </span>
<span class="co">##  [72] cachem_1.0.6                                      </span>
<span class="co">##  [73] cli_3.2.0                                         </span>
<span class="co">##  [74] generics_0.1.2                                    </span>
<span class="co">##  [75] RSQLite_2.2.10                                    </span>
<span class="co">##  [76] evaluate_0.15                                     </span>
<span class="co">##  [77] stringr_1.4.0                                     </span>
<span class="co">##  [78] fastmap_1.1.0                                     </span>
<span class="co">##  [79] yaml_2.3.5                                        </span>
<span class="co">##  [80] ragg_1.2.2                                        </span>
<span class="co">##  [81] knitr_1.37                                        </span>
<span class="co">##  [82] bit64_4.0.5                                       </span>
<span class="co">##  [83] fs_1.5.2                                          </span>
<span class="co">##  [84] beanplot_1.2                                      </span>
<span class="co">##  [85] scrime_1.3.5                                      </span>
<span class="co">##  [86] purrr_0.3.4                                       </span>
<span class="co">##  [87] KEGGREST_1.32.0                                   </span>
<span class="co">##  [88] sparseMatrixStats_1.4.2                           </span>
<span class="co">##  [89] nlme_3.1-152                                      </span>
<span class="co">##  [90] doRNG_1.8.2                                       </span>
<span class="co">##  [91] mime_0.12                                         </span>
<span class="co">##  [92] nor1mix_1.3-0                                     </span>
<span class="co">##  [93] xml2_1.3.3                                        </span>
<span class="co">##  [94] biomaRt_2.48.3                                    </span>
<span class="co">##  [95] compiler_4.1.1                                    </span>
<span class="co">##  [96] rstudioapi_0.13                                   </span>
<span class="co">##  [97] filelock_1.0.2                                    </span>
<span class="co">##  [98] curl_4.3.2                                        </span>
<span class="co">##  [99] png_0.1-7                                         </span>
<span class="co">## [100] interactiveDisplayBase_1.30.0                     </span>
<span class="co">## [101] tibble_3.1.6                                      </span>
<span class="co">## [102] bslib_0.3.1                                       </span>
<span class="co">## [103] stringi_1.7.6                                     </span>
<span class="co">## [104] GenomicFeatures_1.44.2                            </span>
<span class="co">## [105] desc_1.4.0                                        </span>
<span class="co">## [106] minfi_1.38.0                                      </span>
<span class="co">## [107] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0</span>
<span class="co">## [108] lattice_0.20-44                                   </span>
<span class="co">## [109] Matrix_1.3-4                                      </span>
<span class="co">## [110] nloptr_2.0.0                                      </span>
<span class="co">## [111] multtest_2.48.0                                   </span>
<span class="co">## [112] vctrs_0.3.8                                       </span>
<span class="co">## [113] rhdf5filters_1.4.0                                </span>
<span class="co">## [114] pillar_1.7.0                                      </span>
<span class="co">## [115] lifecycle_1.0.1                                   </span>
<span class="co">## [116] BiocManager_1.30.16                               </span>
<span class="co">## [117] jquerylib_0.1.4                                   </span>
<span class="co">## [118] data.table_1.14.2                                 </span>
<span class="co">## [119] bitops_1.0-7                                      </span>
<span class="co">## [120] httpuv_1.6.5                                      </span>
<span class="co">## [121] rtracklayer_1.52.1                                </span>
<span class="co">## [122] GenomicRanges_1.44.0                              </span>
<span class="co">## [123] R6_2.5.1                                          </span>
<span class="co">## [124] BiocIO_1.2.0                                      </span>
<span class="co">## [125] bookdown_0.24                                     </span>
<span class="co">## [126] promises_1.2.0.1                                  </span>
<span class="co">## [127] IRanges_2.26.0                                    </span>
<span class="co">## [128] codetools_0.2-18                                  </span>
<span class="co">## [129] boot_1.3-28                                       </span>
<span class="co">## [130] MASS_7.3-54                                       </span>
<span class="co">## [131] assertthat_0.2.1                                  </span>
<span class="co">## [132] rhdf5_2.36.0                                      </span>
<span class="co">## [133] SummarizedExperiment_1.22.0                       </span>
<span class="co">## [134] openssl_1.4.6                                     </span>
<span class="co">## [135] rprojroot_2.0.2                                   </span>
<span class="co">## [136] rjson_0.2.21                                      </span>
<span class="co">## [137] withr_2.4.3                                       </span>
<span class="co">## [138] GenomicAlignments_1.28.0                          </span>
<span class="co">## [139] Rsamtools_2.8.0                                   </span>
<span class="co">## [140] S4Vectors_0.30.2                                  </span>
<span class="co">## [141] GenomeInfoDbData_1.2.6                            </span>
<span class="co">## [142] hms_1.1.1                                         </span>
<span class="co">## [143] quadprog_1.5-8                                    </span>
<span class="co">## [144] grid_4.1.1                                        </span>
<span class="co">## [145] tidyr_1.2.0                                       </span>
<span class="co">## [146] base64_2.0                                        </span>
<span class="co">## [147] minqa_1.2.4                                       </span>
<span class="co">## [148] DelayedMatrixStats_1.14.3                         </span>
<span class="co">## [149] illuminaio_0.34.0                                 </span>
<span class="co">## [150] MatrixGenerics_1.4.3                              </span>
<span class="co">## [151] numDeriv_2016.8-1.1                               </span>
<span class="co">## [152] Biobase_2.52.0                                    </span>
<span class="co">## [153] shiny_1.7.1                                       </span>
<span class="co">## [154] restfulr_0.0.13</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Fernanda Veitzman, Lissette Gomez, Tiago Silva, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d" class="external-link">Lily Wang</a>, <a href="https://github.com/gabrielodom" class="external-link">Gabriel Odom</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
