<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract contiguous co-methylated genomic regions from a list of
  pre-defined genomic regions — CoMethAllRegions • coMethDMR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract contiguous co-methylated genomic regions from a list of
  pre-defined genomic regions — CoMethAllRegions"><meta property="og:description" content="Extract contiguous co-methylated genomic regions from a list of
  pre-defined genomic regions"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coMethDMR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/vin1_Introduction_to_coMethDMR_geneBasedPipeline.html">Introduction to coMethDMR</a>
    </li>
    <li>
      <a href="../articles/vin2_BiocParallel_for_coMethDMR_geneBasedPipeline.html">Genome-wide methylation analysis using coMethDMR via parallel computing</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/TransBioInfoLab/coMethDMR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract contiguous co-methylated genomic regions from a list of
  pre-defined genomic regions</h1>
    <small class="dont-index">Source: <a href="https://github.com/TransBioInfoLab/coMethDMR/blob/HEAD/R/CoMethAllRegions.R" class="external-link"><code>R/CoMethAllRegions.R</code></a></small>
    <div class="hidden name"><code>CoMethAllRegions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extract contiguous co-methylated genomic regions from a list of
  pre-defined genomic regions</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">CoMethAllRegions</span><span class="op">(</span>
  <span class="va">dnam</span>,
  betaToM <span class="op">=</span> <span class="cn">FALSE</span>,
  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span>, <span class="st">"spearman"</span><span class="op">)</span>,
  rDropThresh_num <span class="op">=</span> <span class="fl">0.4</span>,
  minCpGs <span class="op">=</span> <span class="fl">3</span>,
  genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg19"</span>, <span class="st">"hg38"</span><span class="op">)</span>,
  arrayType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"450k"</span>, <span class="st">"EPIC"</span><span class="op">)</span>,
  <span class="va">CpGs_ls</span>,
  file <span class="op">=</span> <span class="cn">NULL</span>,
  returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>,
  output <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CpGs"</span>, <span class="st">"dataframe"</span><span class="op">)</span>,
  nCores_int <span class="op">=</span> <span class="fl">1L</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>dnam</dt>
<dd><p>matrix (or data frame) of beta values, with row names = CpG IDs,
column names = sample IDs. This is typically genome-wide methylation beta
values.</p></dd>
<dt>betaToM</dt>
<dd><p>indicates if converting methylation beta values to mvalues</p></dd>
<dt>method</dt>
<dd><p>method for computing correlation,
can be "spearman" or "pearson"</p></dd>
<dt>rDropThresh_num</dt>
<dd><p>threshold for min correlation between a cpg with sum
of the rest of the CpGs</p></dd>
<dt>minCpGs</dt>
<dd><p>minimum number of CpGs to be considered a "region".
Only regions with more than <code>minCpGs</code> will be returned.</p></dd>
<dt>genome</dt>
<dd><p>Human genome of reference hg19 or hg38</p></dd>
<dt>arrayType</dt>
<dd><p>Type of array, can be "450k" or "EPIC"</p></dd>
<dt>CpGs_ls</dt>
<dd><p>list where each item is a character vector of CpGs IDs.
This should be CpG probes located closely on the array.</p></dd>
<dt>file</dt>
<dd><p>an RDS file with clusters of CpG locations (i.e. CpGs
located closely to each other on the genome). This file can be generated
by the <code><a href="WriteCloseByAllRegions.html">WriteCloseByAllRegions</a></code> function.</p></dd>
<dt>returnAllCpGs</dt>
<dd><p>When there is not a contiguous comethylated region in
the inputting pre-defined region, <code>returnAllCpGs = TRUE</code> indicates
outputting all the CpGs in the input regions (regardless of statistical
significance), while <code>returnAllCpGs = FALSE</code> indicates not returning
any CpGs not contained in comethylated clusters. Defaults to <code>FALSE</code>,
and we provide this option for debugging purposes only.</p></dd>
<dt>output</dt>
<dd><p>a character vector of CpGs or a dataframe of CpGs along with
rDrop info</p></dd>
<dt>nCores_int</dt>
<dd><p>Number of computing cores to be used when executing code
in parallel. Defaults to 1 (serial computing).</p></dd>
<dt>...</dt>
<dd><p>Dots for additional arguments passed to the cluster constructor.
See <code><a href="CreateParallelWorkers.html">CreateParallelWorkers</a></code> for more information.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>When <code>output = "dataframe"</code> is selected, returns a list of data
   frames, each with <code>CpG</code> (CpG name), <code>Chr</code> (chromosome number),
<code>MAPINFO</code> (genomic position), <code>r_drop</code> (correlation between the
   CpG with rest of the CpGs), <code>keep</code> (indicator for co-methylated CpG),
<code>keep_contiguous</code> (index for contiguous comethylated subregions).
   When <code>output = "CpGs"</code> is selected, returns a list, each item is a
   list of CpGs in the contiguous co-methylated subregion.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>There are two ways to input genomic regions for this function: (1)
   use <code>CpGs_ls</code> argument, or (2) use <code>file</code> argument. Examples of
   these files are in /inst/extdata/ folder of the package.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in">   <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">betasChr22_df</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">CpGisland_ls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span class="r-in">     <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span class="r-in">       <span class="st">"extdata"</span>,</span>
<span class="r-in">       <span class="st">"CpGislandsChr22_ex.rds"</span>,</span>
<span class="r-in">       package <span class="op">=</span> <span class="st">'coMethDMR'</span>,</span>
<span class="r-in">       mustWork <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in">     <span class="op">)</span></span>
<span class="r-in">   <span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">   <span class="va">coMeth_ls</span> <span class="op">&lt;-</span> <span class="fu">CoMethAllRegions</span> <span class="op">(</span></span>
<span class="r-in">     dnam <span class="op">=</span> <span class="va">betasChr22_df</span>,</span>
<span class="r-in">     betaToM <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span class="r-in">     method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span class="r-in">     CpGs_ls <span class="op">=</span> <span class="va">CpGisland_ls</span>,</span>
<span class="r-in">     arrayType <span class="op">=</span> <span class="st">"450k"</span>,</span>
<span class="r-in">     returnAllCpGs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span class="r-in">     output <span class="op">=</span> <span class="st">"CpGs"</span></span>
<span class="r-in">   <span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> snapshotDate(): 2021-05-18</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> see ?sesameData and browseVignettes('sesameData') for documentation</span>
<span class="r-in"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Fernanda Veitzman, Lissette Gomez, Tiago Silva, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d" class="external-link">Lily Wang</a>, <a href="https://github.com/gabrielodom" class="external-link">Gabriel Odom</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

